File: benchmarks/runtime/programs/c/networking.c
/* networking — the equivalent-semantics C peer for the tuonelang networking
 * workload (ADR-0014's socket effects). The identical sequence: per round,
 * create an IPv4 TCP socket listening on an ephemeral loopback port, learn
 * the port with getsockname, connect to it (the backlog completes the
 * handshake before accept), accept, write 8 sixteen-byte chunks (128 bytes)
 * from the client, read them back on the server **one byte per read(2)
 * call** (mirroring the tuonelang program's byte-at-a-time read_byte
 * crossings), and close all three descriptors. The observable result is one
 * round's byte count (reassigned, not accumulated): 128, the exit byte. */
#include <arpa/inet.h>
#include <netinet/in.h>
#include <string.h>
#include <sys/socket.h>
#include <unistd.h>

static long long round_trip(long long chunks) {
    struct sockaddr_in addr;
    socklen_t alen = sizeof(addr);
    int one = 1;
    int listener, client, server;
    long long i, count, want;
    listener = socket(AF_INET, SOCK_STREAM, 0);
    if (listener < 0) return -1;
    setsockopt(listener, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(0);
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    if (bind(listener, (struct sockaddr *)&addr, sizeof(addr)) != 0 ||
        listen(listener, 16) != 0 ||
        getsockname(listener, (struct sockaddr *)&addr, &alen) != 0) {
        close(listener);
        return -2;
    }
    client = socket(AF_INET, SOCK_STREAM, 0);
    if (client < 0) {
        close(listener);
        return -3;
    }
    if (connect(client, (struct sockaddr *)&addr, sizeof(addr)) != 0) {
        close(client);
        close(listener);
        return -3;
    }
    server = accept(listener, 0, 0);
    if (server < 0) {
        close(client);
        close(listener);
        return -4;
    }
    for (i = 0; i < chunks; i++) {
        write(client, "0123456789abcdef", 16);
    }
    count = 0;
    want = chunks * 16;
    while (count < want) {
        unsigned char b;
        if (read(server, &b, 1) != 1) return -5;
        count++;
    }
    close(client);
    close(server);
    close(listener);
    return count;
}

int main(void) {
    long long result = 0;
    long long r;
    for (r = 0; r < 100; r++) {
        result = round_trip(8);
    }
    return (int)(result & 0xff);
}

File: benchmarks/runtime/programs/go/networking.go
// networking — the equivalent-semantics Go peer for the tuonelang networking
// workload (ADR-0014's socket effects). The identical sequence: per round,
// listen on an ephemeral loopback port, dial it (the backlog completes the
// handshake before Accept), accept, write 8 sixteen-byte chunks (128 bytes)
// from the client, read them back on the server **one byte per Read call**
// (mirroring the tuonelang program's byte-at-a-time read_byte crossings and
// the C peer's one-byte read(2) loop), and close all three. The observable
// result is one round's byte count (reassigned, not accumulated): 128, the
// exit byte.
package main

import (
	"net"
	"os"
)

func roundTrip(chunks int64) int64 {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return -1
	}
	client, err := net.Dial("tcp", listener.Addr().String())
	if err != nil {
		listener.Close()
		return -3
	}
	server, err := listener.Accept()
	if err != nil {
		client.Close()
		listener.Close()
		return -4
	}
	chunk := []byte("0123456789abcdef")
	for i := int64(0); i < chunks; i++ {
		client.Write(chunk)
	}
	var count int64 = 0
	want := chunks * 16
	b := make([]byte, 1)
	for count < want {
		n, _ := server.Read(b)
		if n != 1 {
			return -5
		}
		count++
	}
	client.Close()
	server.Close()
	listener.Close()
	return count
}

func main() {
	var result int64 = 0
	for r := int64(0); r < 100; r++ {
		result = roundTrip(8)
	}
	os.Exit(int(result & 0xff))
}
